avr-hal-generic/src/i2c.rs

```diff
@@ -425,51 +514,57 @@ macro_rules! impl_twi_i2c {
-                //    to it. Note that this flag is not automatically cleared by hardware when executing the interrupt routine. Also note
-                //    that clearing this flag starts the operation of the TWI, so all accesses to the TWI Address Register (TWAR), TWI
-                //    Status Register (TWSR), and TWI Data Register (TWDR) must be complete before clearing this flag
+                //    This bit is set by hardware when the TWI has finished its current job and
+                //    expects application software response.  If the I-bit in SREG and TWIE in TWCR
+                //    are set, the MCU will jump to the TWI Interrupt Vector. While the TWINT Flag
+                //    is set, the SCL low period is stretched. The TWINT Flag must be cleared by
+                //    software by writing a logic one to it. Note that this flag is not
+                //    automatically cleared by hardware when executing the interrupt routine. Also
+                //    note that clearing this flag starts the operation of the TWI, so all accesses
+                //    to the TWI Address Register (TWAR), TWI Status Register (TWSR), and TWI Data
+                //    Register (TWDR) must be complete before clearing this flag
                 //
-                // NOTE: In the start() fn above sets the `twint` flag, which `I think` is setting it to true:
+                // NOTE: the start() fn above sets the `twint` flag, setting it to `true`:
                 //    
                 //      .$twint().set_bit()
                 //
-                // But 'true' === 1 which in the language of the spec here means `cleared`...
+                // But in rust, a bool 'true' === 1. It's confusing that the datasheet says a bit
+                // is cleared by writing a logic one to it:
                 //
                 //    The TWINT Flag must be cleared by software by writing a logic one to it.
                 //
-                // Here in avr-device/src/generic.rs
-                //
-                //    impl<FI> FieldReader<bool, FI> {
-                //        /// Value of the field as raw bits.
-                //        #[inline(always)]
-                //        pub fn bit(&self) -> bool {
-                //            self.bits
-                //        }
-                //        /// Returns `true` if the bit is clear (0).
-                //        #[inline(always)]
-                //        pub fn bit_is_clear(&self) -> bool {
-                //            !self.bit()
-                //        }
-                //        /// Returns `true` if the bit is set (1).
-                //        #[inline(always)]
-                //        pub fn bit_is_set(&self) -> bool {
-                //            self.bit()
-                //        }
-                //    }
+                // And the FieldReader also explicitly uses 'clear' to refer to 0 and 'set' to
+                // refer to 1:
                 //
```

Comment:
So basically, I was just trying to understand this weird over-use of the terms `clear` and `set`.

Reply:
This is pretty standard terminology in the bare-metal world :D

Not sure how familiar you are with digital logic, but think about the `TWINT` bit like a logic buffer/flip-flop.  Different signals will do different things to the flip-flop:

- Reading the bit in the register just presents the flip-flop output signal (either high or low = 1 or 0).
- If the CPU writes a 1 to the bit in the register, the flip-flop will be reset -> set to 0 -> it will be _cleared_.
- Certain hardware conditions will set the flip-flop to 1 -> _set_ the bit.
- While the flip-flop is _set_ (logic high on its output), SCL is forced low (no clock pulses).
- On falling edge of the flip-flop output (transition from high to low), TWI will start operation (depending on some other circumstances).

Also see my other comment above about hardware registers in general.

Reply:
Thx!